offsite-activity-signup/Code.gs

```diff
@@ -0,0 +1,346 @@
+ * and current schedule.
+ */
+function makeChoice_(attendee, activitiesById) {
+  for(var i in attendee.preferences) {
+    var activity = activitiesById[attendee.preferences[i]];
+    if (!activity) {
+      continue;
+    }
+    var canJoin = checkAvailability_(attendee, activity);
+    if (canJoin) {
+      attendee.assigned.push(activity);
+      activity.roster.push(attendee);
+      return;
+    }
+  }
+}
+
+/**
+ * Checks that an activity has capacity and doesn't conflict with
+ * previously assigned activities.
+ */
+function checkAvailability_(attendee, activity) {
+  if (activity.capacity <= activity.roster.length) {
+    return false;
+  }
+  var timesConflict = attendee.assigned.some(function(assignedActivity) {
+    return !(assignedActivity.startAt.getTime() > activity.endAt.getTime() || activity.startAt.getTime() > assignedActivity.endAt.getTime());
+  });
+  if (timesConflict) {
+    return false;
+  }
+  return true;
+};
+
+/**
+ * Populates a sheet with the assigned activities for each attendee.
+ */
+function writeAttendeeAssignments_(ss, attendees) {
+  var sheetName = 'Activities by person';
+  var sheet = ss.getSheetByName(sheetName);
+  if (!sheet) {
+    sheet = ss.insertSheet(sheetName);
+  }
+  sheet.clear();
+  sheet.appendRow(["Email address", "Activities"]);
+  sheet.getRange("B1:1").merge();
+  attendees.forEach(function(attendee) {
+    var assignedActivities = attendee.assigned.map(function(activity) {
+      return activity.description;
+    });
+    var row = [attendee.email].concat(assignedActivities);
+    sheet.appendRow(row);
+  });
+  sheet.setFrozenRows(1);
+  sheet.getRange("1:1").setFontWeight("bold");
+  sheet.autoResizeColumns(1, sheet.getLastColumn());
+}
+
+/**
+ * Populates a sheet with the rosters for each activity.
+ */
+function writeActivityRosters_(ss, activities) {
+  var sheetName = 'Activity rosters';
+  var sheet = ss.getSheetByName(sheetName);
+  if (!sheet) {
+    sheet = ss.insertSheet(sheetName);
+  }
+  sheet.clear();
+  var rows = [];
+  // Format each roster as a column in the sheet
+  activities.forEach(function(activity, colIndex) {
+    if (!rows[0]) {
+      rows[0] = fillArray_([], activities.length, '');
+    }
+    rows[0][colIndex] = activity.description;
+    activity.roster.forEach(function(attendee, rowIndex) {
+      if (!rows[rowIndex+1]) {
+        rows[rowIndex+1] = fillArray_([], activities.length, '');
+      }
+      rows[rowIndex+1][colIndex] = attendee.email;
+    });
+  });
+  rows.forEach(function(row) {
+    sheet.appendRow(row);
+  });
+  sheet.setFrozenRows(1);
+  sheet.getRange("1:1").setFontWeight("bold");
+  sheet.autoResizeColumns(1, sheet.getLastColumn());
+}
+
+/**
+ * Loads the activity schedule.
+ */
+function loadActivitySchedule_(ss) {
+  var timeZone = ss.getSpreadsheetTimeZone();
+  var sheet = ss.getSheetByName('Activity Schedule');
+  var rows = sheet.getSheetValues(2, 1, sheet.getLastRow() - 1, sheet.getLastRow());
```

Comment:
If using frozen rows, use the number of frozen rows plus one, instead of hard-coding 2, as the start row.